libraries/AC_Avoidance/AC_Avoid.cpp

```diff
@@ -229,20 +229,25 @@ void AC_Avoid::adjust_velocity_z(float kP, float accel_cmss, float& climb_rate_c
     // limit acceleration
     const float accel_cmss_limited = MIN(accel_cmss, AC_AVOID_ACCEL_CMSS_MAX);
 
-    bool limit_alt = false;
-    float alt_diff = 0.0f;   // distance from altitude limit to vehicle in metres (positive means vehicle is below limit)
+    bool limit_max_alt = false;
+    float alt_max_diff = 0.0f;   // distance from max altitude limit to vehicle in metres (positive means vehicle is below ceiling limit)
 
     const AP_AHRS &_ahrs = AP::ahrs();
 
-    // calculate distance below fence
-    AC_Fence *fence = AP::fence();
-    if ((_enabled & AC_AVOID_STOP_AT_FENCE) > 0 && fence && (fence->get_enabled_fences() & AC_FENCE_TYPE_ALT_MAX) > 0) {
-        // calculate distance from vehicle to safe altitude
+    // Vertical fence limits
+    if ((_enabled & AC_AVOID_STOP_AT_FENCE) > 0) {
+        AC_Fence *fence = AP::fence();
         float veh_alt;
+        // TODO: shouldn't this be relative to origin instead, to be consistent with the code 24 lines bellow?
         _ahrs.get_relative_position_D_home(veh_alt);
-        // _fence.get_safe_alt_max() is UP, veh_alt is DOWN:
-        alt_diff = fence->get_safe_alt_max() + veh_alt;
-        limit_alt = true;
+
+        // calculate distance below ceiling fence
+        if (fence && (fence->get_enabled_fences() & AC_FENCE_TYPE_ALT_MAX) > 0) {
```

Comment:
this part ins't good as in case we don't have fence or it is disable, we will now waste time on get the z pos.